### PR TITLE
feat(Query): Add cacheControl directive that adds headers in the response

### PR DESCRIPTION
**Pull request by anurags92**
### Motivation
This PR adds support for users to enable browser/CDN side caching for their queries. Users can specify `@cacheControl(maxAge: 15)` at the top of their query. This would result in the response being returned with appropriate HTTP headers so that browsers and CDNs can cache the response.
More details [can be read on discuss](https://discuss.dgraph.io/t/rfc-persistent-queries-and-cached-results/10636).

```
query @cacheControl(maxAge: 15){
  queryReview(filter: { comment: {alloftext: "Fantastic"}}) {
    comment
    by {
      username
    }
    about {
      name
    }
  }
}
```
http headers returned:
```
Cache-Control: public,max-age=15
Vary: Accept-Encoding
```

This PR fixes: GRAPHQL-722

### Components affected by this PR 

1. Directives
2. Http responses 

### Does this PR break backwards compatibility?

No

### Testing

Added following tests in:

[WIP]

### Fixes 
Fixes GRAPHQL-722 

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6799)

### CODE REVIEW

File: graphql/schema/wrappers.go [hunk 1]
```diff
@@ -343,6 +344,13 @@ func (o *operation) Mutations() (ms []Mutation) {
 	return
 }
 
+func (o *operation) CacheControl() string {
+	if o.op.Directives.ForName(cacheControlDirective) == nil {
+		return ""
```
Comment (gja) [inline graphql/schema/wrappers.go:349]: private,no-cache would work instead of empty string

File: graphql/schema/wrappers.go [hunk 2]
```diff
@@ -343,6 +344,13 @@ func (o *operation) Mutations() (ms []Mutation) {
...
+	}
+	return "public,max-age=" + o.op.Directives.ForName(cacheControlDirective).Arguments[0].Value.Raw
```
Comment (gja) [inline graphql/schema/wrappers.go:351]: probably read the argument by name instead of arguments[0]?

File: graphql/resolve/resolver.go [hunk 3]
```diff
@@ -484,6 +484,10 @@ func (r *RequestResolver) Resolve(ctx context.Context, gqlReq *schema.Request) *
 	// we can just execute it.
 	switch {
 	case op.IsQuery():
+		resp.Header = make(map[string][]string)
+		if len(op.CacheControl()) != 0 {
+			resp.Header.Set(schema.CacheControlHeader, op.CacheControl())
```
Comment (gja) [inline graphql/resolve/resolver.go:489]: don't forget to add  ^^^. 
`Vary: Accept-Encoding`.

You always have to add this header when putting Cache-Control as per the http spec

**Review by gja — APPROVED** [review]

File: graphql/schema/wrappers.go (hunk 2 above)
Comment (anurags92) [reply graphql/schema/wrappers.go:351]: The name `maxAge` is inside the 0'th index. 
Screenshot1603793557

File: graphql/schema/wrappers.go (hunk 1 above)
Comment (anurags92) [reply graphql/schema/wrappers.go:349]: adding header only if the directive is there, otherwise no header.

**Review by abhimanyusinghgaur — APPROVED** [review]
:lgtm: 

> Reviewed 3 of 5 files at r1, 1 of 2 files at r2, 3 of 46 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6799#-:-MKdo2EW2tTQLrT96mbn:bg7igkv)* status: 7 of 52 files reviewed, 3 unresolved discussions (waiting on @abhimanyusinghgaur, @anurags92, @gja, @MichaelJCompton, and @pawanrawal)

**Pull request merged**